[Loc] fix test failures in German locale

MaStr11 opened the pull request:
Fixes #48532

MaStr11 commented on src/EditorFeatures/CSharpTest/EmbeddedLanguages/RegularExpressions/CSharpRegexParserTests.cs:
```diff
@@ -55,8 +55,11 @@ private static SyntaxToken GetStringToken(string text)
+            var actual = TreeToText(sourceText, tree)
+                .Replace("\"", DoubleQuoteEscaping)
+                .Replace("&quot;", DoubleQuoteEscaping);
+            Assert.Equal(expected.Replace("\"", DoubleQuoteEscaping), actual);
```
The german translation uses double quotes in the message.
https://github.com/dotnet/roslyn/blob/e48c86563b48bc290d4dd90fd348bda2ac86d915/src/Features/Core/Portable/xlf/FeaturesResources.de.xlf#L1913-L1917

MaStr11 commented on src/Features/Core/Portable/xlf/FeaturesResources.de.xlf:
```diff
@@ -642,7 +642,7 @@ Stellen Sie sicher, dass Sie den Bezeichner "tt" für Sprachen verwenden, für d
       <trans-unit id="Modifying_source_file_will_prevent_the_debug_session_from_continuing_due_to_internal_error">
         <source>Modifying source file '{0}' will prevent the debug session from continuing due to internal error: {1}</source>
-        <target state="needs-review-translation">Durch das Ändern der Quelldatei "{0}" wird verhindert, dass die Debugsitzung aufgrund eines internen Fehlers fortgesetzt wird: {1}.</target>
+        <target state="needs-review-translation">Durch das Ändern der Quelldatei "{0}" wird verhindert, dass die Debugsitzung aufgrund eines internen Fehlers fortgesetzt wird: {1}</target>
```
The test is green after this change if the msbuild property `EnableXlfLocalization` is set to true.

Youssef1313 replied:
I don't think XLF files should only be manually modified by a localization team member.

MaStr11 replied:
The loc team needs to be in the loop for sure. If they want me to revert the change I can do so easily.

> A better approach IMO is to use the English version of resources in the tests.

Not in this case because the failing test revealed a wrong translation. The dot is superfluous because you end up with `!.` at the end of the sentence.

`ENC0080: Durch das Ändern der Quelldatei \"src.cs\" wird verhindert, dass die Debugsitzung aufgrund eines internen Fehlers fortgesetzt wird: System.NullReferenceException: NullRef!.`

Youssef1313 replied:
There was a discussion somewhere before on a similar issue and (if I recall correctly) the loc team said some languages require different punctuation or something of that kind.

MaStr11 replied:
@JoeRobich My explanation of what is going on here:

The ` {1}` is replaced by a formatted exception in the form: `ExcpetionType: ExceptionMessage`. *ExceptionMessage* includes a sentence closing punctuation mark (at least in the BCL). The German translation adds a superfluous dot punctuation mark.

Youssef1313 reviewed (commented):
A better approach IMO is to use the English version of resources in the tests.
This is done by using LocalizableResourceString.ToString(CultureInfo.InvariantCulture) of something of that kind.

MaStr11 commented:
No. This is not possible/desired and the comppiler team is explicit forcing their tests to be runable in arbitrary cultures. It was a big effort to get there for the whole solution:


https://github.com/dotnet/roslyn/issues/23837#issuecomment-359860799

That's is also why one of the CI builds runs under a Spanish locale.

JoeRobich commented:
@MaStr11 Thanks for submitting a fix.  

@wli3 @cristianosuzuki77 Can you sign off that the XLF change here is OK?

JoeRobich reviewed (changes requested):
Blocking until we get signoff from the LOC team.

wli3 commented:
@cristianosuzuki77 could you ask the loc team to have the string changed in their database without the ending dot?

cristianosuzuki77 commented:
The comments above about the double quotes and full-stop marker for German is correct. 
German language requires these modifications.

cristianosuzuki77 commented:
For the comment about the rendered phrase having a "!" : this punctuation must be in the source string too. Not all languages have the concept of "!".

MaStr11 commented:
@cristianosuzuki77

> For the comment about the rendered phrase having a "!" : this punctuation must be in the source string too. 

The final message reads :`ENC0080: Durch das Ändern der Quelldatei \"src.cs\" wird verhindert, dass die Debugsitzung aufgrund eines internen Fehlers fortgesetzt wird: System.NullReferenceException: NullRef!.`

`System.NullReferenceException: NullRef!` is somethng like `exception.GetType().Name + ": " + exception.Message`

`exception.Message` includes punctuation. In the test that fails it happens to be "NullRef!" and that is the source of the exclamation mark.

cristianosuzuki77 commented:
@MaStr11 Could the second placeholder be inside quotes as well, to signify it is a text of itself?
Modifying source file '{0}' will prevent the debug session from continuing due to internal error: '{1}'

MaStr11 commented:
Quotes are fine too (But in German we use double quotes in such occasions. Single quotes are rare and are only used for quotations inside quotations.)

cristianosuzuki77 commented:
@MaStr11 This then would be the change:
Source in English:  Modifying source file '{0}' will prevent the debug session from continuing due to internal error: '{1}'
German would be: Durch das Ändern des Texts von "{0}" wird die Fortsetzung der Debugsitzung aufgrund des folgenden internen Fehlers verhindert: "{1}".

Example of it in runtime: `ENC0080: Durch das Ändern der Quelldatei \"src.cs\" wird verhindert, dass die Debugsitzung aufgrund eines internen Fehlers fortgesetzt wird: System.NullReferenceException: "NullRef!".`

MaStr11 commented:
@cristianosuzuki77 Do you want me to change the `src/Features/Core/Portable/xlf/FeaturesResources.de.xlf`? Or will this be overridden in your loc passes anyway?

cristianosuzuki77 commented:
This needs to be added from your side. Then the loc tools will detect the update and will proceed with the translation.
Note: this will be applicable to all other languages too. The last placeholder will be in seen as in quotes in English and translated accordingly.

MaStr11 commented:
@cristianosuzuki77 I did the changes as requested. I hope I got it right. Please note:

* I also changed `Modifying_body_of_member_will_prevent_the_debug_session_from_continuing_due_to_internal_error` now, because it is comparable to `Modifying_source_file_will_prevent_the_debug_session_from_continuing_due_to_internal_error`
* I removed the dot from de, es, and pl
* I just copied the quotes from {0} to {1} for each language.

The `test.cmd` is currently running on my PC and CI builds are not finished yet. I will report back later for the test results.

cristianosuzuki77 commented:
@MaStr11 Be advised that translations may still be required to add a full stop.  To reduce this chance, could you add the example in the <note> element when the '{1}' could have the '!' point and refrain if possible to add a full-stop marker in the translation? This will orient the translators to consider twice adding back the full-stop in the translation.

MaStr11 commented:
This comment needs to be in the resx, doesn't it?

cristianosuzuki77 commented:
Correct. Once you put in there it will be extract into the xliff.

MaStr11 commented:
> Quotes are fine too

I have to retract that statement. Here is why:
The test was verifying a diagnostics id and message. This particular message also contained an exception. An exception is formatted as
* The exception type (e.g. `System.NullReferenceException`)
* ": "
* The exception message (e.g. `"Nullref!"`)
* A NewLine
* The Stacktrace

The full sample is (with the new double quotation marks):
```
Durch das Ändern der Quelldatei "src.cs" wird die Fortsetzung der Debugsitzung aufgrund des folgenden internen Fehlers verhindert: "System.NullReferenceException: NullRef!
   bei Microsoft.CodeAnalysis.CSharp.EditAndContinue.UnitTests.CSharpEditAndContinueAnalyzerTests.<>c__DisplayClass18_0.<AnalyzeDocumentAsync_InternalError>b__0(SyntaxNode node) in C:\IFP\OS\roslyn\src\EditorFeatures\CSharpTest\EditAndContinue\CSharpEditAndContinueAnalyzerTests.cs:Zeile 741.
   bei Microsoft.CodeAnalysis.EditAndContinue.AbstractEditAndContinueAnalyzer.<AnalyzeDocumentAsync>d__72.MoveNext() in C:\IFP\OS\roslyn\src\Features\Core\Portable\EditAndContinue\AbstractEditAndContinueAnalyzer.cs:Zeile 419."
```
The stack trace contains the line number (in debug builds) and is therefore not stable. That's why the logic in the test was cutting of the whole message after the first NewLine (but in a faulty way, which led to the test failing as reported in  #48532).

https://github.com/dotnet/roslyn/blob/793e757500b311071d09ecf92af967a45090d5f2/src/EditorFeatures/CSharpTest/EditAndContinue/CSharpEditAndContinueAnalyzerTests.cs#L737-L755

This bug was fixed in 41036e26, so the test failure is no longer reproducible with the latest master.

With all this in mind, there are still some other issues to fix here:
* I need to revert the quotes around the `{1}`. As you can see above, the quotation marks are not very helpful.
* We still need to remove the `.` after `{1}` in German (and some other languages). Otherwise, the user will see two dots at the end of the message (the last line above would look like so: `... \AbstractEditAndContinueAnalyzer.cs:Zeile 419..` ⬅️ see the double dots here)
* The Turkish translation places the `{1}` inside the message. This does not play well with the multiline exception message including the stacktrace. The `{1}` should be placed at the end (I can't do this change as I'm not speaking Turkish).

I will do the first two points, but the Turkish translation needs to be fixed on your side.

MaStr11 commented:
@cristianosuzuki77 I did the changes as described in https://github.com/dotnet/roslyn/pull/49111#issuecomment-737911791:

I reverted the quotes (3e8c595) and then removed the dot from de, es, and pl (b84fa5c). I also added a note about the pitfalls of the `{1}` parameter to resx (d71410c) and xlf (b71f336).

The Turkish translation needs still to be fixed, though.

cristianosuzuki77 commented:
The intent of the quotes within the second placeholder was to let the end-use see that the information within it is as complete external sentence. 
@MaStr11 Sounds good. This will be fixed on the loc side once you place the comments asking the translation to abstain from placing a full-stop if the grammar of the languages mandates it. The loc system will then request a review for this change and the new translation will be returned. 
Even for the German, Spanish and Polish files there's the need for the translation to be redone in the loc side otherwise they will re-appear once the file is updated by a new translation/fix.
You also will need to comment about the relative position of the placeholders. EG as in case of {1}, it must not be placed at the middle of the translation.

MaStr11 commented:
> You also will need to comment about the relative position of the placeholders. EG as in case of {1}, it must not be placed at the middle of the translation.

I did so in d71410c638b94d054029ae855daa4d2139d3a738

```XML
<comment>{1} is a multi-line exception message including a stacktrace. Place it at the end of the message and don’t add any punctation after or around {1}</comment>
```

Is there anything else I need to do?

cristianosuzuki77 commented:
The translation should be coming 5 days after the merge.

MaStr11 commented:
@JoeRobich As far as I can tell, this PR can be merged. A sign off by @CyrusNajmabadi for the changes in `CSharpRegexParserTests.cs` might still be needed.
